substrate/frame/asset-rewards/src/lib.rs

```diff
@@ -0,0 +1,718 @@
+			pool_info.total_tokens_staked.saturating_accrue(amount);
+			Pools::<T>::insert(pool_id, pool_info);
+
+			// Update PoolStakers.
+			staker_info.amount.saturating_accrue(amount);
+			PoolStakers::<T>::insert(pool_id, &caller, staker_info);
+
+			// Emit event.
+			Self::deposit_event(Event::Staked { who: caller, pool_id, amount });
+
+			Ok(())
+		}
+
+		/// Unstake tokens from a pool.
+		#[pallet::call_index(2)]
+		pub fn unstake(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let caller = ensure_signed(origin)?;
+
+			// Always start by updating the pool rewards.
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			let staker_info = PoolStakers::<T>::get(pool_id, &caller).unwrap_or_default();
+			let (mut pool_info, mut staker_info) =
+				Self::update_pool_and_staker_rewards(&pool_info, &staker_info)?;
+
+			// Check the staker has enough staked tokens.
+			ensure!(staker_info.amount >= amount, Error::<T>::NotEnoughTokens);
+
+			// Unfreeze staker assets.
+			// TODO: (blocked https://github.com/paritytech/polkadot-sdk/issues/3342)
+
+			// Update Pools.
+			pool_info.total_tokens_staked.saturating_reduce(amount);
+			Pools::<T>::insert(pool_id, pool_info);
+
+			// Update PoolStakers.
+			staker_info.amount.saturating_reduce(amount);
+			PoolStakers::<T>::insert(pool_id, &caller, staker_info);
+
+			// Emit event.
+			Self::deposit_event(Event::Unstaked { who: caller, pool_id, amount });
+
+			Ok(())
+		}
+
+		/// Harvest unclaimed pool rewards for a staker.
+		#[pallet::call_index(3)]
+		pub fn harvest_rewards(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			staker: Option<T::AccountId>,
+		) -> DispatchResult {
+			let caller = ensure_signed(origin)?;
+
+			let staker = staker.unwrap_or(caller.clone());
+
+			// Always start by updating the pool and staker rewards.
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			let staker_info =
+				PoolStakers::<T>::get(pool_id, &staker).ok_or(Error::<T>::NonExistentStaker)?;
+			let (pool_info, mut staker_info) =
+				Self::update_pool_and_staker_rewards(&pool_info, &staker_info)?;
+
+			// Transfer unclaimed rewards from the pool to the staker.
+			let pool_account_id = Self::pool_account_id(&pool_id)?;
+			T::Assets::transfer(
+				pool_info.reward_asset_id,
+				&pool_account_id,
+				&staker,
+				staker_info.rewards,
+				Preservation::Preserve,
+			)?;
+
+			// Emit event.
+			Self::deposit_event(Event::RewardsHarvested {
+				who: caller.clone(),
+				staker,
+				pool_id,
+				amount: staker_info.rewards,
+			});
+
+			// Reset staker rewards.
+			staker_info.rewards = 0u32.into();
+			PoolStakers::<T>::insert(pool_id, &caller, staker_info);
+
+			Ok(())
+		}
+
+		/// Modify a pool reward rate.
+		#[pallet::call_index(4)]
+		pub fn set_pool_reward_rate_per_block(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			new_reward_rate_per_block: T::Balance,
```

Comment:
When does this take effect? Does it take effect for all stakers immediately, in retrospect of when they next interact with the pool (which could be unfair)?

Reply:
It effect all stakers immediately, but only for their rewards going forward since the pool rewards are changed :)

It's tested in the integration test and `staker_rewards_are_affected_correctly` test if you wanted to check those.